MVP

jasper7466 opened the pull request:
## Что нового

- Свёрстаны основные компоненты в соответствии с макетом
- Реализован локальный мини-сервер авторизации
- Настроен обмен post-сообщениями между окнами
- Подключено NgRx-хранилище
- Ручное добавление адресата по номеру телефона и ИНН
- Загрузка файлов по кнопке (file-browser), перетаскиванием (drag'n'drop) и конвертация в base64
- Конвертация из base64, скачивание, отображение в pdfPreviewer'е
- Работа с json-хранилищем в текстовом поле карточки сделки
- Создание/удаление пакета документов, сохранение и чтение его Id из карточки crm-системы
- Экраны статусов пакета документов, поллинг информации о пакетах
- API v1 и в необходимом объёме v2
- Обход CORS-блокировки посредством прозрачного postMessage-http-прокси
- Перехват и отображение http-ошибок (в т.ч. через проксированный postMessageHttpClient)

jasper7466 committed:
Deploy. Настроена публикация на gh-pages

jasper7466 committed:
Вёрстка. Temp. commit

jasper7466 committed:
Tools. Реализован локальный сервер авторизации для отладки

jasper7466 committed:
Tools. Auth-dev-server fix

jasper7466 committed:
Tools. Auth-dev-server fix

jasper7466 committed:
Notification page/service. Добавлена страница уведомлений

jasper7466 committed:
Landing page. Добавлена страница для редиректа при запуске вне CRM

jasper7466 committed:
PostMessageService. Реализован сервис обмена post-сообщениями

jasper7466 committed:
CrmService. Реализован сервис post-запросов к виджету-обёртке

jasper7466 committed:
Routing. Подключен роутинг

jasper7466 committed:
Startup page / Init. Добавлена страница загрузки и логика инициализации

jasper7466 committed:
Fix. AuthDevServer. Изменён pingURI

jasper7466 committed:
Remove. AccountPlateComponent. Удалён компонент смены аккаунта

jasper7466 committed:
Fix. WidgetPage. Исправлен скроллинг

jasper7466 committed:
DevProxy. Настроен конфиг прокси для angular-dev-сервера

jasper7466 committed:
NgRx. Подключен стейт-менеджмент

jasper7466 committed:
Addressee. Реализовано ручное добавление адресата

jasper7466 committed:
Files. Реализовано добавление файлов через DnD + FileBrowser

jasper7466 committed:
Files. Загрузка в хранилище, конвертация, связка с UI

jasper7466 committed:
Documents. Создание пакета документов

jasper7466 committed:
Documents. Удаление пакета документов, обработка статусов до подписания

jasper7466 committed:
Readme. Добавлено описание проекта

jawsspot commented on src/app/components/blocks/main-content/addressee/add-addressee-form/add-addressee-form.component.ts:
```diff
@@ -0,0 +1,82 @@
+
+  ngOnInit(): void {}
+
+  changeSearchSelector(selector: SearchSelector) {
```
Проставь модификатопры для всех метоодов в каждом классе,
Типизируй все методы, не хватает возращаемого типа

jawsspot commented on src/app/components/blocks/main-content/addressee/add-addressee-form/add-addressee-form.component.ts:
```diff
@@ -0,0 +1,82 @@
+export class AddAddresseeFormComponent implements OnInit {
+  @Output() onSubmit = new EventEmitter<void>();
+
+  searchSelector: SearchSelector = 'vatId';
```
модификаторы

jawsspot commented on src/app/components/blocks/main-content/addressee/add-addressee-form/add-addressee-form.component.ts:
```diff
@@ -0,0 +1,82 @@
+          vatId: this.value,
+        })
+      );
+      return;
```
для чего этот return?

jawsspot commented on src/app/components/blocks/main-content/addressee/add-addressee-form/add-addressee-form.component.ts:
```diff
@@ -0,0 +1,82 @@
+        })
+      );
+
+      this.nopaperApiService.checkByPhone$(this.value).subscribe((response) => {
```
нужно отписываться по дестрою компоненты

jawsspot commented on src/app/components/blocks/main-content/documents/documents-drop-area/documents-drop-area.component.html:
```diff
@@ -0,0 +1,27 @@
+      id="fileInputId"
+      #fileInputRef
+      appFileBrowser
+      (onChanged)="filesService.filesHandler($event)"
```
нельзя использовать сервисы напрямую в разметке, оберни в метод класса этого темплейта, дай понятное имя и дергай его

jawsspot commented on src/app/components/blocks/main-content/stage-header/status-banner/status-banner.component.ts:
```diff
@@ -0,0 +1,15 @@
+  @Input() status: status = '';
+  constructor() {}
+
+  ngOnInit(): void {}
```
не нужно реализовывать ngOnInit там где он не используется

jawsspot commented on src/app/services/api/amo/amo-api.service.ts:
```diff
@@ -0,0 +1,86 @@
+    });
+  }
+
+  private get$<T>(path: string): Observable<T> {
```
финскую нотацию используй только для полей класса, но не методов, это общепринятая практика

jawsspot commented on src/app/services/api/amo/amo-api.service.ts:
```diff
@@ -0,0 +1,86 @@
+    });
+  }
+
+  get getCompaniesCustomFieldsAll$(): Observable<customField[]> {
```
`СustomField`

jawsspot commented on src/app/services/api/amo/amo-api.service.ts:
```diff
@@ -0,0 +1,86 @@
+    );
+  }
+
+  patchLeadById(id: number, data: Partial<PatchLeadRequestBody>) {
```
всегда типизируй возвращаемое значение, так код становится более самодокументируемым

jawsspot commented on src/app/services/api/amo/amo-api.types.ts:
```diff
@@ -0,0 +1,147 @@
+import { Nullable } from './../../../../types/common';
+
+type customFieldType =
```
типы всегда с большой буквы

jawsspot commented on src/app/services/api/amo/amo-api.types.ts:
```diff
@@ -0,0 +1,147 @@
+  };
+};
+
+export interface GetCustomFieldsResponse {
```
еще бестпрактис - писать I в названии интерфейса `IGetCustomFieldsResponse`

jawsspot commented on src/app/services/crm.service.ts:
```diff
@@ -0,0 +1,112 @@
+  getPacketId$() {
+    return this.amoApiService.getLeadById$(this.leadId).pipe(
+      map((lead) => lead.custom_fields_values),
+      map((fields) =>
```
не понял зачем столько операторов мап, 
лучше сделай в одном логическом блоке, но используй понятные переменные

jawsspot commented on src/app/services/files.service.ts:
```diff
@@ -0,0 +1,64 @@
+    const fileRecords: FileRecord[] = [];
+
+    for (const file of Array.from(files)) {
+      const id = this.counter;
```
для чего ?

jawsspot commented on src/app/services/nopaper.service.ts:
```diff
@@ -0,0 +1,100 @@
+    private nopaperApiService: NopaperApiService,
+    private crmService: CrmService
+  ) {
+    this.store.select(packetIdSelector).subscribe((id) => (this.packetId = id));
```
не забывай про отписки, тем более в конструкторе

jawsspot commented on src/app/services/nopaper.service.ts:
```diff
@@ -0,0 +1,100 @@
+    this.store
+      .select(addresseeSelector)
+      .pipe(
+        tap((addressee) => {
```
вообще не понятно, что происходит в этом пайплайне до строки с методом postDraft$ 
вынеси этот мапинг в отдельный мметод, назовие его так, чтоб можно было понять зачем это все и не лезть внутрь

jawsspot commented on src/app/components/blocks/header/status-bar/status-bar.component.ts:
```diff
@@ -0,0 +1,15 @@
+
+  constructor() { }
+
+  ngOnInit(): void {
```
во всём проекте овер много мест, по этому дальнейшие коментарии  применимы ко всем файлам и все требуют исправлений

jawsspot commented on src/app/app.module.ts:
```diff
@@ -1,16 +1,113 @@
 @NgModule({
   declarations: [
-    AppComponent
+    AppComponent,
```
декоратор получился очень громоздким, а структура проекта трудно читаема,
к примеру - компоненты папки common можно объединить в один модуль.
так же давай избавимся от глубокой вложенности компонент.  
к примеру document-count-label - это самостоятельный компонент, но он лежит гдет-то глубоко в папке header - так нельзя.  самостоятельные(реюзабельные) компоненты выноси в common, не надо группировать по месту нахождения на странице

jawsspot replied:
и не делай глубокую вложенность папок с компонентами, храни плоским массивом

jasper7466 committed:
Добавлено отображение статусов пакета/документов
- Дерево компонентов развёрнуто в плоскую структуру

jasper7466 committed:
Pipes. Добавлено преобразование stepName -> StatusLabelStatus

jasper7466 committed:
Routes. Изменена структура роутинга и хранилища
- Добавлена страница списка пакетов
- Добавлен поллинг пакетов

jasper7466 committed:
Misc. Доработка структуры роутов, сервисов и др.

jasper7466 committed:
NPM. Audit fix

jasper7466 committed:
GitIgnore. Добавлена директория для локальных npm-пакетов

jasper7466 committed:
API. Nopaper. PacketInfo. Добавлено получение информации о пакете

jasper7466 committed:
Temp. commit

jasper7466 committed:
API. Nopaper. Переход на внешний сервер авторизации

jasper7466 committed:
Temp. commit

jasper7466 committed:
API. Nopaper. PacketDetails. Fix 404

jasper7466 committed:
NPM. Audit fix

jasper7466 committed:
FS. Atomic Design

jasper7466 committed:
Pipes. Declension. Добавлено склонение числительных

jasper7466 committed:
Refactor. Modal template

jasper7466 committed:
UI. PacketsList. Стилизован список пакетов документов

jasper7466 committed:
Refactor. Documents: List, FoldableList, ListItem, Uploader

jasper7466 committed:
Tools. Auth-dev-server fix

jasper7466 committed:
Store. Удалена неиспользуемая часть хранилища

jasper7466 committed:
Misc. Переход в статус "prepare"
- Реализована смена статуса "new" -> "prepare"
- Добавлены дополнительные роуты
- Прочие сопутствующие исправления

jasper7466 committed:
UI. Draft banner

jasper7466 committed:
PreviewPage. Подготовка к интеграции pdf-превьювера

jasper7466 committed:
Utils. GetContentType. Получение MIME-типа файла по его расширению

jasper7466 committed:
Utils. Base64ToFile. Конвертация из "укороченной" base64-строки в объект типа "File"

jasper7466 committed:
PreviewPage. Встроен pdf-превьювер

jasper7466 committed:
StatusLabel. Добавлены новые статусы

jasper7466 committed:
SenderSignPage. Добавлен роут для этапа ожидания подписи

jasper7466 committed:
BugFix. NopaperService. Добавлена отписка от селектора хранилища в цепочке switchMap'ов

jasper7466 committed:
BugFix. Minor fixes

jasper7466 committed:
Refactor. Modules. Компоненты, пайпы и директивы вынесены в "SharedModules"

jasper7466 committed:
Refactor. Modules. ServicesCoreModule

jasper7466 committed:
Fix. Set packet title

jasper7466 committed:
Tools. Auth-dev-server fix

jasper7466 committed:
Refactor. Api
- Добавлен базовый API-сервис, от которого наследуются частные API-сервисы
- Получение amo-токена вынесено в отдельный api-сервис
- Провайд всех api-сервисов вынесен в ServicesCoreModule
- Дописаны аннотации для методов - Исправлен нейминг типов/интерфейсов

jasper7466 committed:
Refactor. CommonService
- Часть логики вынесена в общий сервис
- Частично добавлено документирование методов
- Незначительные исправления

jasper7466 committed:
PacketPageEnd. Восстановление после рефакторинга
- Добавлен компонент "sign-info-item"
- Добавлены селекторы хранилща для групп документов
- Восстановлена страница "packet-page-end"
- Правки в компонент модального окна

jasper7466 committed:
PacketPageReceiverPreview. Реализован соответствующий роут

jasper7466 committed:
Framework. Update @angular/cli @angular/core

jasper7466 committed:
Framework. Update @ngrx/store

jasper7466 committed:
PacketPage. Fix. Исправлена ошибка навигирования "назад" по событию "popstate"

jasper7466 committed:
CodeStyle. ESLint. Подключен линтинг

jasper7466 committed:
DocumentsListItem. Реализовано скачивание файлов

jasper7466 committed:
UI. DocumentsListItem. Fix

jasper7466 committed:
Environments. Обновлены baseUrl

jasper7466 committed:
Routing/NavBar. Реализована панель навигации
- Переконфигурирован роутинг
- Панель навигации строится автоматически от корня до текущего роута (при наличии title)
- Поправлены "особенности" навигации браузерными инструментами на странице widget-page-packet

jasper7466 committed:
WidgetPageNew. Fix. Очистка введённых данных по нажатию "Отмена"

jasper7466 committed:
API. Добавлен рест для генерации ссылки на подписание

jasper7466 committed:
UI. Fix. PacketListItem. Title overflow
- Поправлено отображение длинных названий пакетов (обрезка + троеточие)

jasper7466 committed:
ShareLink. Реализована кнопка для получения КЭП-ссылки

jasper7466 committed:
API. amoCRM. Добавлен рест получения прикреплённых к сделке файлов

jasper7466 committed:
Temp. commit

jasper7466 committed:
Temp. commit

jasper7466 committed:
API. AccessTokenLocalApiService
Добавлено:
- отдельный API-сервис для получения токена через локальный сервер
- переменная окружения для выбора API получения токена
- фабрика для провайдинга AccessTokenApiService

jasper7466 committed:
API. Base URL. Fix

jasper7466 committed:
API. Nopaper. Refactor
- Эндпоинты вынесены в отдельные файлы
- Добавлен слой адаптеров для эндпоинтов

jasper7466 committed:
API. NopaperV2
- Создан сервис
- Реализованы несколько эндпоинтов

jasper7466 committed:
Store. Refactor
- Temp. commit

jasper7466 committed:
Interfaces. Вынесены общие интерфейсы

jasper7466 committed:
Temp. commit

jasper7466 committed:
API/Store/Components. Fix TS errors

jasper7466 committed:
PostMessageTransport. Refactor
- Транспорт через window.postMessage вынесен в отдельный сервис
- Работа с post-сообщениями вынесена в отдельный api-like-сервис

jasper7466 committed:
MockServices.
- WindowService
- WindowMockService
- AmoPostApiMockService

jasper7466 committed:
Startup. Fix. Реинициализация приложения при обновлении страницы

jasper7466 committed:
API. CRM. Refactor
- Эндпоинты вынесены в отдельные файлы
- Добавлен слой адаптеров для эндпоинтов

jasper7466 committed:
Transport. PostMessage. Fix types

jasper7466 committed:
CrmTokenService

jasper7466 committed:
CrmJsonStorageService
- Работа с текстовым хранилищем вынесена в отдельный сервис

jasper7466 committed:
Temp. commit. Misc. refactor/fixes

jasper7466 committed:
PostMessage http proxy. Temp. commit

jasper7466 committed:
PostMessage http proxy. Temp. commit

jasper7466 committed:
PostMessage http proxy. Temp. commit

jasper7466 committed:
PostMessage http proxy. Temp. commit
- TDD

jasper7466 committed:
PostMessage http proxy. Temp. commit
- PostMessageXhr DI test

jasper7466 committed:
NPM. Audit fix

jasper7466 committed:
PostMessage http proxy. Temp. commit

jasper7466 committed:
PostMessage http proxy. Минимально необходимая реализация

jasper7466 committed:
PostMessage http proxy. Реализован функционал метода "abort()"
- Отправка post-сообщения при прерывании запроса (если уже инициирован)
- Добавлены отписки при успешном получении ответа\ошибке

jasper7466 committed:
API. Host-change + fix

jasper7466 committed:
Store. BugFix. Packets reducer

jasper7466 committed:
BugFixes

jasper7466 committed:
Error handling. Проверка активации/конфигурации виджета

jasper7466 committed:
Error handling. Добавлена обработка http-ошибок

jasper7466 committed:
ReviewFix

jasper7466 committed:
API. Ручное изменение шага пакета, отзыв пакета
- В APIv1 отключены/ограничены более не поддерживаемые эндпоинты
- В APIv2 добавлен эндпоинт для отзыва пакета документов

jasper7466 committed:
Logic. Отзыв/удаление/подписание пакета
- Отзыв пакета переключен на APIv2
- Возможность удаления пакета заменена на его отзыв
- Отключена возможность "подписать и отправить"

jasper7466 committed:
ESLint. Fix

daleunixal commented on src/app/pages/main-pages/landing-page/landing-page.component.ts:
```diff
@@ -0,0 +1,15 @@
+  styleUrls: ['./landing-page.component.css'],
+})
+export class LandingPageComponent implements OnInit {
+  constructor(private routingService: RoutingService) {}
```
Перенос строки + приватные поля начинаются с земли. F/E: _routingService

jasper7466 replied:
Про перенос строки - не совсем понятно
Префикс "_" для приватных полей - исправлено
F/E: - что это означает? ))

daleunixal commented on src/app/pages/main-pages/main-pages-shared.module.ts:
```diff
@@ -0,0 +1,25 @@
+    StartupPageComponent,
+    WidgetPageComponent,
+  ],
+  imports: [CommonModule, ComponentsSharedModule, RouterModule],
```
Без инлайнов, запятая в конце

jasper7466 replied:
исправлено

daleunixal commented on src/app/app-routing.module.ts:
```diff
@@ -0,0 +1,71 @@
+import { PacketPagePrepareComponent } from './pages/packet-pages/packet-page-prepare/packet-page-prepare.component';
+
+export const routes: Routes = [
+  {
```
Indent 4 space на всем проекте

jasper7466 replied:
исправлено

daleunixal commented on src/app/components/atoms/atoms-shared.module.ts:
```diff
@@ -0,0 +1,52 @@
+    IconDownloadComponent,
+    IconDocumentComponent,
+  ],
+  imports: [CommonModule, TemplatesSharedModule, PipesSharedModule],
```
no-inline

jasper7466 replied:
исправлено

daleunixal commented on src/app/components/atoms/button/button.component.ts:
```diff
@@ -0,0 +1,26 @@
+  styleUrls: ['./button.component.css'],
+})
+export class ButtonComponent {
+  @Input() styleOption: ButtonStyleOption = 'solid';
```
Модификатор доступа. После декоратора перенос строки.

jasper7466 replied:
модификатор доступа - исправлено везде
декоратор над полем - исправлено везде

daleunixal commented on src/app/components/atoms/input/input.component.ts:
```diff
@@ -0,0 +1,24 @@
+  @Input() maskPrefix = '';
+
+  @Output()
+  keyUp = new EventEmitter<string>();
```
Модификатор доступа везде

jasper7466 replied:
исправлено

daleunixal commented on src/app/components/molecules/documents-drop-area/documents-drop-area.component.ts:
```diff
@@ -0,0 +1,31 @@
+export class DocumentsDropAreaComponent {
+  private leadId$ = this.store.select(leadIdSelector);
+
+  constructor(private filesService: FilesService, private store: Store) {}
```
no-inline

daleunixal replied:
Приватные поля начинаются с земли

jasper7466 replied:
приватные поля с "_" - исправлено

jasper7466 replied:
про `no-inline` - не совсем понятно

если это к аргументам, то вроде по код-гайду 2 инлайн-аргумента допустимо, про`max-len` ничего не сказано (у меня сейчас 90, было 80)

если к фигурным скобкам - в гайде конкретного пункта нету по этому моменту, можно только судить по некоторым примерам, причём там местами по-разному приведено (просто newline и newline + пустая строка):

``` JavaScript
public joinCompanies(): void {
}
```

``` JavaScript
constructor(
    private _a: any,
    private _b: any,
    private _c: any, //
) {
}

constructor(private _a: any) {

}
```

с ходу не нашёл подходящих правил ни для линтера, ни для преттиера

daleunixal commented on src/app/components/molecules/documents-drop-area/documents-drop-area.component.ts:
```diff
@@ -0,0 +1,31 @@
+        take(1),
+        filter((leadId) => typeof leadId === 'number'),
+        map((leadId) => leadId as number)
+        // switchMap((leadId) => this.crmService.getLeadAttachments(leadId))
```
?

jasper7466 replied:
это заготовка будущего (?) функционала
закомментировал в коде и вёрстке, открыл тикет, добавил TODO

daleunixal commented on src/app/components/molecules/documents-drop-area/documents-drop-area.component.ts:
```diff
@@ -0,0 +1,31 @@
+  }
+
+  protected loadFromLeadButtonHandler(): void {
+    this.leadId$
```
Если переключение на поток из сервиса не нужен, то этот метод не вносит никакого импакта

jasper7466 replied:
это заготовка будущего (?) функционала
закомментировал в коде и вёрстке, открыл тикет, добавил TODO

daleunixal commented on src/app/components/molecules/documents-list-item/documents-list-item.component.ts:
```diff
@@ -0,0 +1,52 @@
+      this.nopaperApiService
+        .getFilesByIds([this.fileId])
+        .pipe(take(1))
+        .subscribe((files) => {
```
Данная запись является deprecated. См. https://rxjs.dev/guide/observable

jasper7466 replied:
не совсем понятно...

имеется ввиду, что нужно подписываться в таком виде?

``` JavaScript
observable.subscribe({
  next(x) {
    // ...
  },
  error(err) {
    // ...
  },
  complete() {
    // ...
  },
});
```

но там в примерах много и аналогичных подписок:

``` JavaScript
foo.subscribe((x) => {
  console.log(x);
});
foo.subscribe((y) => {
  console.log(y);
});
```

daleunixal commented on src/app/components/organisms/addressee-selector/addressee-selector.component.ts:
```diff
@@ -0,0 +1,37 @@
+export class AddresseeSelectorComponent {
+  @ViewChild(ModalComponent) modal: ModalComponent | undefined;
+
+  protected openModalEmitter: EventEmitter<void> = new EventEmitter<void>();
```
Описание куда идет emit

jasper7466 replied:
непонятно))
это должен быть просто произвольный текстовый комментарий? или есть какой-то устоявшийся формат?

daleunixal commented on src/app/components/organisms/modal-sign-info/modal-sign-info.component.ts:
```diff
@@ -0,0 +1,32 @@
+
+  protected signatures$ = this.store.select(signatureSelector);
+
+  protected closeTrigger = new EventEmitter<void>();
```
Описание куда идет emit

jasper7466 replied:
непонятно))
это должен быть просто произвольный текстовый комментарий? или есть какой-то устоявшийся формат?

daleunixal commented on src/app/directives/drag-and-drop.directive.ts:
```diff
@@ -0,0 +1,45 @@
+  }
+  @Output() fileDropped = new EventEmitter<FileList>();
+
+  fileOver = false;
```
Модификатор

jasper7466 replied:
исправлено

daleunixal commented on src/app/pages/packet-pages/packet-page-end/packet-page-end.component.ts:
```diff
@@ -0,0 +1,66 @@
+
+  ngOnInit(): void {
+    console.log(this.router.url);
+    const id = this.route.parent?.snapshot.paramMap.get('id');
```
Лучше асинхронно из this._activatedRoute.paramMap. Если это прокидывается в query использовать аналогичный метод из activatedRoute

jasper7466 replied:
исправлено везде

daleunixal commented on src/app/pages/packet-pages/packet-page-receiver-preview/packet-page-receiver-preview.component.ts:
```diff
@@ -0,0 +1,67 @@
+  ) {}
+
+  ngOnInit(): void {
+    const id = this.route.parent?.snapshot.paramMap.get('id');
```
Лучше асинхронно

jasper7466 replied:
исправлено везде

daleunixal commented on src/app/services/api/amo-api/amo-api.module.ts:
```diff
@@ -0,0 +1,46 @@
+  'POST MESSAGE HTTP CLIENT TOKEN'
+);
+
+const PostMessageHttpClientFactory = (
```
Опиши, чем обусловлен кастомный флоу отправки запроса.

jasper7466 replied:
Использование модифицированного http-клиента обусловлено невозможностью выполнения прямых http-запросов к amoCRM-API из SPA-виджета на клиенте с другого источника (блокируются политиками CORS).

Одно из изначальных требований: реализовать всё по-максимуму на клиенте, без необходимости делать дополнительный backend.

Единственный видимый лайфхак - делегировать http-запросы виджету-обёртке, который живёт на своём домашнем домене и не подвержен блокировке CORS.

В данном случае подменяется xhr-фабрика, которая отдаёт вместо нативного браузерного объекта XMLHttpRequest - модифицированный. Он в достаточной степени эмулирует функционал нативного объекта, но сетевого запроса не делает. Вместо этого происходит postMessage-запрос к виджету-обёртке в определённом формате.

Виджет-обёртка обрабатывает эти сообщения, формирует и выполняет сетевые запросы. Обратными сообщениями возвращается тело ответа/ошибки, прогресс выполнения запроса и прочий контекст.

Для клиентского кода такая подмена выглядит незаметной, можно писать методы запросов к API как обычно. Интерцепторы тоже срабатывают.

Добавил соответствующий JSDoc-комментарий к фабрике.

daleunixal commented on src/app/services/api/amo-api/amo-api.service.ts:
```diff
@@ -0,0 +1,88 @@
+const BASE_URL_COMPILER: (domain: string) => string = environment.getAmoBaseUrl;
+
+@Injectable({
+  providedIn: AmoApiModule,
```
Лучше провайд указать непосредственно в модуле, чем в декораторе.

jasper7466 replied:
Изначально указывал провайды в модулях, но потом наткнулся на примечание в доках, что будто бы предпочтительнее делать это в декораторе. Написано (https://angular.io/guide/providers) следующее:

``` JavaScript
import { Injectable } from '@angular/core';
import { UserModule } from './user.module';

@Injectable({
  providedIn: UserModule,
})
export class UserService {
}
```
"The example above shows the preferred way to provide a service in a module. This method is preferred because it enables tree-shaking of the service if nothing injects it. ..."

daleunixal commented on src/app/services/api/amo-post-api-mock/amo-post-api-mock.service.ts:
```diff
@@ -0,0 +1,22 @@
+      isAdminUser: true,
+      isWidgetActive: true,
+      isWidgetConfigured: true,
+    }).pipe(delay(2000));
```
Для чего тут delay?

jasper7466 replied:
Это моковый сервис, в продакшене не используется.
Имитирует postMessage-ответ от виджета-обёртки и позволяет запускать основной SPA-виджет вне CRM-системы.
Задержка в пайпе - для имитации задержки при выполнении запроса, чтобы экран инициализации не переключался слишком быстро.

daleunixal commented on src/app/services/api/amo-post-api/amo-post-api.service.ts:
```diff
@@ -0,0 +1,21 @@
+  ) {}
+
+  public getCrmContext = () => {
+    return getCrmContextEndpoint.call(this);
```
Call в angular скорее путает, чем упрощает жизнь. Лучше добавить метод для запроса нормальный

jasper7466 replied:
Почему юзаю `call`:

Изначально внутри приложения завязался на форматы ответа API (не лучшее решение, теперь уже очевидно).
Позже вышло API v2, которое слабо соответствует v1, нужно было перейти на него (перешёл частично, т.к. некоторая совместимость с v1 всё-таки реализована на стороне сервера). Перепиливать всё приложение не очень хотелось.

Решил попытаться изолировать внутренние интерфейсы "ядра" приложения от внешних, изобразить что-то вроде "луковичной" архитектуры: контракт задаёт ядро, а внешние слои (API) содержат в себе адаптеры и подстраиваются под ядро. Тогда можно подсовывать любые версии API-сервиса без изменения ядра приложения, главное снабдить их методы адаптерами.

Все методы запросов к одному API у меня выделены в свой сервис. Теперь в каждом пайпе каждого метода сервиса нужно было бы разместить адаптер тела запроса и адаптер тела ответа. Они иногда довольно громоздкие, из-за чего код методов вытягивается в трудночитаемую партянку, и сам сервис тоже сильно много строк занимает.

Решил эту проблему выделением кода методов запроса в отдельные файлы-эндпоинтов: `name.endpoint.ts`, сложил в отдельную директорию на одном уровне с ts-файлом сервиса. В них разместил описание интерфейсов запроса/ответа, адаптеры запроса/ответа, и функцию запроса. Всё, что используется вместе и больше нигде не нужно - лежит компактно в одном файле и не засоряет автокомплит нигде больше не использующимися экспортами.

Функцию запроса экспортирую и использую в коде сервиса, а `call` там нужен для того, чтобы эта функция получила нужный контекст сервиса, т.к. дёргает методы `httpClient`'а не напрямую, а через обёртки базового `ApiService`'а (в этих обёртках подставляются заголовки, чтобы не дублировать их каждый раз в каждом методе запроса, подставляется базовый URL и ещё там можно кастомно обрабатывать ошибки для конкретно этого Api-класса).

Таким образом ts-файл API-сервиса является своего рода "баррелем" для методов запроса, описанных в отдельных функциях и разложенных по своим файлам.

Короче, если избавлюсь от `call` и запихну всю обработку в сервис - он сильно раздуется.

Вероятно, тут опять что-то не то с архитектурой) Буду рад любым советам/рекомендациям/ссылкам на референсы как сделать лучше или как оно у вас принято))

daleunixal commented on src/app/services/api/api.service.ts:
```diff
@@ -0,0 +1,64 @@
+
+export class ApiService {
+  private _baseUrl: string;
+  private headers = new HttpHeaders();
```
Очень странное. Лучше прокидывать хэдеры именно в запрос, а не формировать их подобным образом.

jasper7466 replied:
Не хотел одни и те же заголовки прокидывать в каждом методе, т.к. они всегда одинаковы (в пределах класса конкретного API). Поэтому спрятал это дело в базовый API-класс, в нём обёртки над методами запросов httpClient'а подставляют эти заголовки и в наследуемых классах конкретных API уже не нужно этого делать

daleunixal commented on src/app/services/services-core.module.ts:
```diff
@@ -0,0 +1,19 @@
+  providers: [CommonLogicService],
+})
+export class ServicesCoreModule {
+  constructor(@Optional() @SkipSelf() parentModule: ServicesCoreModule) {
```
Для чего эта проверка? Есть места где создается локальный sigleton?

jasper7466 replied:
На самом деле я сам слабо понимаю. В прошлом ревью было требование разбить всё на модули. В контексте того, как у меня построено приложение - не очень разобрался как это правильно сделать. Где-то нашёл примеры, что бывают "shared-модули" и "core-модули", и что в последних вроде как неплохо бы делать такую проверку. Если это тут не нужно - уберу

daleunixal reviewed (changes requested):
Было бы отлично убрать все call.
Не хватает земли у приватных параметров/полей и т.д.
Хотелось бы вынести каждый тип/интерфейс в отдельные файлы, в аналогично названные папки, с использование баррелей.
В некоторых местах не хватает комментариев.
Хотелось бы узнать почему использован Atomic Design Pattern
Нужен пост-ревью после правок

daleunixal commented:
Нужно переделать прилагу под наш код-стайл с использованием общей "кухни". Документацию скину в ЛС.

jasper7466 committed:
CodeStyle. Fix. 1.2.5: Приватные свойства должны начинаться с символа подчёркивания

jasper7466 committed:
ESLint. Отключено правило "@typescript-eslint/no-empty-interface"

jasper7466 committed:
ESLint. CodeGuide 1.2.2: Именование классов в "PascalCase"

jasper7466 committed:
ESLint. CodeGuide 1.2.3: Поля класса должны иметь модификатор доступа

jasper7466 committed:
CodeStyle. Fix. 1.2.3: Поля класса должны иметь модификатор доступа

jasper7466 committed:
ESLint. CodeGuide 1.2.5: Приватные свойства должны начинаться с символа подчёркивания

jasper7466 committed:
ESLint. CodeGuide 1.2.4: Именование public/protected-свойств классов в "camelCase"

jasper7466 committed:
ESLint. CodeGuide 1.2.6: Именование методов классов в "camelCase"

jasper7466 committed:
CodeStyle. Fix. 1.2.6: Именование методов классов в "camelCase"

jasper7466 committed:
ESLint. CodeGuide 1.2.7 (2/2): Параметры методов класса и аргументы функций
- (1/2) Именование в "camelCase"
- (2/2) Обязательная типизация

jasper7466 committed:
CodeStyle. Fix. 1.2.7: Именование/типизация аргументов методов/функций

jasper7466 committed:
ESLint. CodeGuide 4.5: Обязательная типизация возвращаемого значения функций/методов

jasper7466 committed:
CodeStyle. Fix. 4.5: Обязательная типизация возвращаемого значения функций/методов

jasper7466 committed:
ESLint. Unspecified: Именование переменных/констант в "camelCase"

jasper7466 committed:
Codestyle. Fix. Unspecified: Именование переменных/констант в "camelCase"

jasper7466 committed:
ESLint. CodeGuide 1.2.8: Не использовать сокращения в именовании сущностей и файлов

jasper7466 committed:
CodeStyle. Fix. 1.2.8: Не использовать сокращения в именовании сущностей и файлов

jasper7466 committed:
ESLint. Unspecified: Проверка орфографии в исходниках (*.ts)

jasper7466 committed:
CodeStyle. Fix. Unspecified: Проверка орфографии в исходниках (*.ts)

jasper7466 committed:
ESLint. CodeGuide 2.3: Именование перечислений в единственном числе

jasper7466 committed:
Docs. Обновлена документация

jasper7466 committed:
AuthDevServer. Fix

jasper7466 committed:
ESLint. CodeGuide 1.2.10 (1): Именование составляющих названия интерфейса в единственном числе

jasper7466 committed:
CodeStyle. Fix. CodeGuide 1.2.10 (1): Именование составляющих названия интерфейса в единственном числе

jasper7466 committed:
ESLint. CodeGuide 1.2.10 (2): Именование интерфейсов с префиксом "I"

jasper7466 committed:
ESLint. CodeGuide 1.2.12: Именовать все асинхронные сущности в "финской" нотации

jasper7466 committed:
CodeStyle. Fix. 1.2.12: Именовать все асинхронные сущности в "финской" нотации

jasper7466 committed:
ESLint. CodeGuide 2.1: Каждое перечисление должно быть определено в отдельном файле

jasper7466 committed:
ESLint. CodeGuide 3.1: Каждый класс должен располагаться в отдельном файле

jasper7466 committed:
ESLint. CodeGuide 3.2: Каждый интерфейс должен быть определён в отдельном файле

jasper7466 committed:
ESLint. CodeGuide 4.6: Не использовать упрощённые конструкции if/else/while

jasper7466 committed:
ESLint/Prettier. CodeGuide 6.1: Отступ: 4 пробела

jasper7466 committed:
CodeStyle. Fix. 6.1: Отступ: 4 пробела

jasper7466 committed:
ESLint. CodeGuide 5.1: Декоратор поля класса на отдельной строке

jasper7466 committed:
CodeStyle. Fix. 5.1: Декоратор поля класса на отдельной строке

jasper7466 committed:
CodeStyle. Fix. CodeGuide 7.1 (1, 2); CodeGuide 3.4 (2)
- Элементы массивов в декораторе модулей должны быть с новой строки (no-inline)
- Висящая запятая в массивах декораторов модулей
- Висящая запятая в списке аргументов методов/конструкторов

jasper7466 committed:
ESLint/Prettier. CodeGuide 7.1 (1,2); CodeGuide 3.4 (2)
- Элементы массивов в декораторе модулей должны быть с новой строки (no-inline)
- Висящая запятая в массивах декораторов модулей
- Висящая запятая в списке аргументов методов/конструкторов

jasper7466 committed:
ReviewFix. Неиспользуемый метод подгрузки файлов из сделки

jasper7466 committed:
ESLint. RxJs recommended

jasper7466 committed:
CodeStyle. Fix. RxJs recommended

jasper7466 committed:
ReviewFix. Асинхронное обращение к параметрам роута

jasper7466 committed:
Readme. Update

jasper7466 committed:
RxJS. Recommended-fix. Fix

jasper7466 committed:
ReviewFix. Документирование кастомного postMessageHttpClient

jasper7466 commented:
@daleunixal, спасибо за ревью)

> Было бы отлично убрать все call.

Отписался в комментариях, мне нужен совет, как тут лучше поступить.

> Не хватает земли у приватных параметров/полей и т.д.

Настроил линтер, везде поправил

> Хотелось бы вынести каждый тип/интерфейс в отдельные файлы, в аналогично названные папки, с использование баррелей.

Настроил линтер, но пока не правил. Вероятно тут в целом беда с архитектурой. Надо понять какой она должна быть и отталкиваться уже от этого.

> В некоторых местах не хватает комментариев.

JSDoc-документирование выполнено частично. Постараюсь покрыть всё, как описано в гайде, но позже.

> Хотелось бы узнать почему использован Atomic Design Pattern

Да какой-то особой причины нет... Требований по архитектуре изначально не поступало.

Сначала был только макет, сверстал, структуру сделал что-то наподобие bem-nested. Общие компоненты вытащил в директорию "common". Вложенность получилась дикая и этим реально было не очень удобно пользоваться.

В предыдущем предварительном ревью порекомендовали хранить всё плоско. Длинный список компонентов в одной директории мне не понравился: когда и кнопка и страница и ещё мелочь какая лежит с чем-то большим рядом.
Однажды довелось использовать AtomicDesign. Мне он показался неплохим. Какой-то компромисс между плоским хранением и сильной вложенностью: помогает разделить примерно по масштабу/смыслу компоненты.

Если нужно - переделаю, но желательно скинуть мне хоть какой-то референс, на что опираться.

> Нужен пост-ревью после правок

К пост-ревью пока не готов. Поправил ещё не всё, но везде в комментах к замечаниям отписался, есть некоторые непонятки.

Прорабатываю код-гайд, что сделано - [отмечено тут](https://github.com/jasper7466/amopaper-widget/issues/2)
Не все требования удаётся легко покрыть правилами... И пока не до конца подружил eslint с prettier'ом, поэтому местами всё ещё есть неконсистентные варианты когда.

daleunixal commented on src/app/services/common-logic.service.ts:
```diff
@@ -0,0 +1,123 @@
+                    this._crmService.checkWidgetStatus();
+                    this._routingService.goPacketsListPage();
+                },
+                error: (error: unknown) => console.log(error),
```
Надо бы уведомлять пользователя о любых ошибках, которые возникают

daleunixal commented on src/app/pages/packet-pages/packet-page-preview/packet-page-preview.component.ts:
```diff
@@ -0,0 +1,81 @@
+
+    protected submitButtonHandler(): void {
+        this.isAwaiting = true;
+        this._commonLogicService.submitPreview$(this._packetId).pipe(take(1)).subscribe();
```
Выкинет необработанную ошибку, это ок?

daleunixal commented on src/app/pipes/pipes-shared.module.ts:
```diff
@@ -0,0 +1,22 @@
+import { NgModule } from '@angular/core';
```
Пайпы лучше убрать по личным папкам. В итоге, относительно каталога, должен торчать только модуль

daleunixal commented on src/app/pages/widget-pages/widget-page-new/widget-page-new.component.ts:
```diff
@@ -0,0 +1,89 @@
+    templateUrl: './widget-page-new.component.html',
+    styleUrls: ['./widget-page-new.component.css'],
+})
+export class WidgetPageNewComponent implements OnInit, OnDestroy {
```
Наименование не сходится с содержанием. Файл назван не согласно стилю. Я бы предложил DocumentCreatePage, document-create.page.ts

daleunixal commented on src/app/pages/widget-pages/widget-page-new/widget-page-new.component.ts:
```diff
@@ -0,0 +1,89 @@
+                combineLatestWith(this.isAllFilesLoaded$),
+                takeUntil(this._onDestroyEmitter),
+            )
+            .subscribe(([isAddresseeAdded, isAllFilesLoaded]) => {
```
isAddressee? Орфография, нужно переименовать.

daleunixal commented on src/app/pages/widget-pages/widget-page-new/widget-page-new.component.ts:
```diff
@@ -0,0 +1,89 @@
+    );
+
+    protected isControlsEnabled = false;
+    protected isAwaiting = false;
```
isAwaiting - чего? Нужно дополнить комментарием, либо переименовать, чтобы наименование отвечала конкретно на вопрос, чего ждем.

daleunixal reviewed (changes requested):
Остальное скину в ЛС

jawsspot commented on src/app/services/api/amo-api/endpoints/get-entity-custom-fields-all.endpoint.ts:
```diff
@@ -0,0 +1,36 @@
+};
+
+export function getEntityCustomFieldsAllEndpoint$(
+    this: ApiService,
```
это очень не очевидная штука и плохая практика, эта функция должна быть методом сервиса `AmoPostApiService` 
нельзя использовать call тогда, когда это не нужно

jawsspot commented on src/app/services/api/amo-post-api/amo-post-api.service.ts:
```diff
@@ -0,0 +1,21 @@
+    ) {}
+
+    public getCrmContext$ = (): Observable<ICrmContext> => {
+        return getCrmContextEndpoint$.call(this);
```
перенести метод сюда

jawsspot commented on src/app/services/api/amo-post-api/amo-post-api.service.ts:
```diff
@@ -0,0 +1,21 @@
+        protected postMessageTransportService: PostMessageTransportService,
+    ) {}
+
+    public getCrmContext$ = (): Observable<ICrmContext> => {
```
реализация должна быть такой же как в access-token-local-api.service.ts

jawsspot commented on src/app/services/api/nopaper-api/nopaper-api.service.ts:
```diff
@@ -0,0 +1,43 @@
+    public getFilesByIds = getFilesByIdsEndpoint$;
+    public getFileSignatures = getFileSignaturesEndpoint$;
+    public getShareLink = getShareLinkEndpoint$;
+    public setPacketStepName = setPacketStepNameEndpoint$;
```
тут должно быть не присвоение, а объявление метода, все функции вынесенные в отдельные файлы перенести сюда и сделать методами класса

jawsspot replied:
тут править не обязательно, но надо кинуть метку deprecated на сервис

jawsspot commented on src/app/services/sub-services/routing.service.ts:
```diff
@@ -0,0 +1,128 @@
+};
+
+@Injectable()
+export class RoutingService {
```
в этом сервисе почти нет реюзабельных методов, а значит - сервис нужно удалить, а методы определять в местах их использования

jawsspot commented on src/app/services/sub-services/crm.service.ts:
```diff
@@ -0,0 +1,115 @@
+@Injectable()
+export class CrmService {
+    private _storagePollingBreaker = new EventEmitter<void>();
+    private _context: ReturnType<typeof crmContextSelector>;
```
удалитьь

jawsspot commented on src/app/services/sub-services/crm.service.ts:
```diff
@@ -0,0 +1,115 @@
+    private _storagePollingBreaker = new EventEmitter<void>();
+    private _context: ReturnType<typeof crmContextSelector>;
+
+    private _context$ = this._store$.select(crmContextSelector);
```
удалить

jawsspot commented on src/app/services/sub-services/crm.service.ts:
```diff
@@ -0,0 +1,115 @@
+        this._context$.subscribe((context) => (this._context = context));
+    }
+
+    public checkWidgetStatus(): void | never {
```
метод можно сделать асинхронным, пусть возвращает Observable

jawsspot commented on src/app/services/sub-services/crm.service.ts:
```diff
@@ -0,0 +1,115 @@
+                );
+            } else {
+                this._notificationService.notify(
+                    notifications.widgetNotConfiguredAdmin,
```
вероятно что тут опечатка

jawsspot commented on src/app/services/sub-services/notification.service.ts:
```diff
@@ -0,0 +1,40 @@
+    constructor(private _routingService: RoutingService) {}
+
+    public notify(context: TNotification): void {
+        this.context = { ...baseContext, ...context };
```
редирект и патч контекста не должены лежать вместе в методе notify
такую логику нужно разделять, а редиректы стараться ВСЕгда юзать в компонентах, в сервисах их трудно найти. Редирект должен быть результатом выполнения какой-то логики, а не частью логики

jawsspot commented on src/app/pages/packet-pages/packet-page-preview/packet-page-preview.component.ts:
```diff
@@ -0,0 +1,81 @@
+            this._packetId = parseInt(id);
+
+            this.packet$ = this._store$.select(packetSelector(this._packetId));
+            this._commonLogicService.getPacketFiles$(this._packetId).subscribe();
```
отписка

jawsspot commented on src/app/services/api/nopaper-api-v2/nopaper-api-v2.service.ts:
```diff
@@ -0,0 +1,29 @@
+
+    public postPacket = createPacketEndpoint$;
+    public getPacketDetails = getPacketDetailsEndpoint$;
+    public revokePacket = revokePacketEndpoint$;
```
тут должно быть не присвоение, а объявление метода, все функции вынесенные в отдельные файлы перенести сюда и сделать методами класса

jawsspot replied:
аналогично ко всем датасервисам
